controllers/controllers.go

```diff
@@ -281,3 +296,55 @@ func markAsFreeOnSuccess(c client.Client, watchNamespace string, recorder record
 	}
 	return err
 }
+
+// markNodeAsUpgrading marks the given node as upgrading by adding an annotation to it. If the number of nodes
+// performing upgrades in parallel exceeds the maximum allowed, an error is returned.
+func markNodeAsUpgrading(ctx context.Context, c client.Client, node *core.Node) error {
+	controllerLocker.Lock()
+	defer controllerLocker.Unlock()
+
+	upgradingNodes, err := findUpgradingNodes(ctx, c, node)
+	if err != nil {
+		return err
+	}
+	if len(upgradingNodes) == 1 && upgradingNodes[0].Name == node.Name {
```

Comment:
we should only try to access property of upgradingNodes if we can ensure no of upgradingNodes are greater than 0.

Reply:
@mansikulkarni96 I don't fully understand your comment. Do you anticipate any issues?

Here the code is treating a special case of ` findUpgradingNodes` as documented in the function' [comment](https://github.com/openshift/windows-machine-config-operator/pull/1901/files#diff-46c1e02dfda1ba21d1ccabdf287921220f29dfdd6a75639fbf8a1a95bce01782R327) where a single-element collection is returned if the given node is found.

Reply:
The scenario where  no node is found and an empty list is returned, is the happy path and is covered below.  All Windows nodes are available for upgrade and it is a first-come first-serve approach to mark the node as upgrading, usually (at least in my local tests) the machinecontroller reacts faster and mark the machine instances first, the byoh instances fall behind.